smarts/core/local_agent.py

```diff
@@ -0,0 +1,52 @@
+# Copyright (C) 2020. Huawei Technologies Co., Ltd. All rights reserved.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+from concurrent import futures
+
+from smarts.core.buffer_agent import BufferAgent
+from smarts.zoo.agent_spec import AgentSpec
+
+
+class LocalAgent(BufferAgent):
+    """A remotely controlled agent."""
+    
+    def __init__(self, act_executor: futures.Executor):
+        self._agent = None
+        self._agent_spec = None
+        self._act_executor = act_executor
+    
+    def act(self, obs):
+        """Call the agent's act function asynchronously and return a Future."""
+        def obtain_future(obs):
+            adapted_obs = self._agent_spec.observation_adapter(obs)
```

Comment:
Why does this need to be done in the future?

Reply:
The intention was to still do multiprocessing for social agents.

Reply:
Here I meant specifically line 37, applying the observation adapter.  I.e., I think the observation adapter could be applied immediately to `obs` and then `adapted_obs` could be passed with the `obtain_future` method to `submit()` on line 44.

But it's not a big deal either way.  I can see a case for defering the application of the adapter as well, so I think it's fine to leave it as is.